### DISCUSSION TITLE
Load Local File

### DISCUSSION THREAD

**Discussion by AdriFernan12** [google/model-viewer#3680] [Q&A]
Hi! I'm using Android Studio and i'm trying to render a 3d object on the modelviewer, when i put a web url it works fine, but what i want to do is a offline app and it doesn't work. I imported the 3d object to my assets folder but when it loads it shows an error:
Fetch API cannot load file:///android_asset/planetLowPoly.glb. URL scheme "file" is not supported.
Is there any way to load a local file? Thanks in advance

**Comment by elalish** [comment]
This is probably a better question for the Android Studio team. I would expect a simple relative path would suffice? MV's GLB should be no different than say, serving an image into a WebView, so maybe look at an example like that.

**Reply by AdriFernan12** [reply]
In this same repo i've seen people saying that is not possible lo load a local file, only executing a local web server, is it possible?

**Reply by nicolas-daures** [reply]
@AdriFernan12 Fetch does not support the `file://` protocol. You can override `fetch` to use `xhr` instead. This is not a problem of ModelViewer but of Three.js which is only based on fetch.

**Comment by luknow93** [comment]
Could anyone help me solve this?
I have `.glb` files in apps assets folder that I was hoping to display offline with the use of MV by using WebView. But no matter what I do they just refuse to load with `file:///` scheme. 
Truth be told, I'm not even sure where to start to try to make it work.

**Reply by elalish** [reply]
I'm not very familiar with WebViews - there's probably a simple way to make this work. Have you tried with local images? As a workaround, you might just fetch the file - you can wrap a Blob up as an ObjectURL and set that as the src.

**Reply by nicolas-daures** [reply]
You can override the fetch function to use xhr in the case of the file protocol:

```
window.fetch = async (...args) => {
    const [request, config] = args;
    let response;
    if (request.url?.startsWith('file://'))
    {
        response = new Promise((resolve, reject) =>
        {
            const xhr = new XMLHttpRequest();
            const url = decodeURI(request.url);
            xhr.open(request.method, url, true);
            xhr.responseType = 'arraybuffer';

            xhr.addEventListener('load', () =>
            {
                resolve({
                    status: xhr.status,
                    url: url,
                    body: xhr.response,
                    arrayBuffer: () => xhr.response
                });
            }, false);
            xhr.addEventListener('error', (error) => reject(error), false);
            xhr.send();
        });
    }
    else
    {
        response = await originalFetch(request, config);
    }
    return response;
};
```

**Reply by luknow93** [reply]
@nicolas-daures 
Hey man!

Awesome thanks for the snippet!

I added it to to my code and while it works im running into the following issue...
`THREE.GLTFLoader: Couldn't load texture blob:file:///a6584161...`
...after which the model loads, it animates but it is indeed stripped of any textures, any idea what could be causing it?

I would try to fix it myself, or at least try to tweak this and that, but my JS knowledge is limited to literally what I learned since yesterday 🙈 I'd really appreciate if you're able to help me fix that or at least try to explain what should I look for when working on finding the fix myself 🙏 

I've also managed to develop secondary solution based on the discussion you guys had in August which boils down to setting up local server on the device that just serves the files from assets back to the WebView. While working fine it sure is slower(load time) than what your snippet is capable of.
